lib/execution.go

```diff
@@ -0,0 +1,562 @@
+	endTime := atomic.LoadInt64(es.endTime)
+	pausedDuration := es.totalPausedDuration
+	es.pauseStateLock.RUnlock()
+
+	if endTime == 0 {
+		pauseTime := atomic.LoadInt64(es.currentPauseTime)
+		if pauseTime != 0 {
+			endTime = pauseTime
+		} else {
+			// The test isn't paused or finished, use the current time instead
+			endTime = time.Now().UnixNano()
+		}
+	}
+
+	return time.Duration(endTime-startTime) - pausedDuration
+}
+
+// Pause pauses the current execution. It acquires the lock, writes
+// the current timestamp in currentPauseTime, and makes a new
+// channel for resumeNotify.
+// Pause can return an error if the test was already paused.
+func (es *ExecutionState) Pause() error {
+	es.pauseStateLock.Lock()
+	defer es.pauseStateLock.Unlock()
+
+	if !atomic.CompareAndSwapInt64(es.currentPauseTime, 0, time.Now().UnixNano()) {
+		return errors.New("test execution was already paused")
+	}
+	es.resumeNotify = make(chan struct{})
+	return nil
+}
+
+// Resume unpauses the test execution. Unless the test wasn't
+// yet started, it calculates the duration between now and
+// the old currentPauseTime and adds it to
+// Resume will emit an error if the test wasn't paused.
+func (es *ExecutionState) Resume() error {
+	es.pauseStateLock.Lock()
+	defer es.pauseStateLock.Unlock()
+
+	currentPausedTime := atomic.SwapInt64(es.currentPauseTime, 0)
+	if currentPausedTime == 0 {
+		return errors.New("test execution wasn't paused")
+	}
+
+	// Check that it's not the pause before execution actually starts
+	if atomic.LoadInt64(es.startTime) != 0 {
+		es.totalPausedDuration += time.Duration(time.Now().UnixNano() - currentPausedTime)
+	}
+
+	close(es.resumeNotify)
+
+	return nil
+}
+
+// ResumeNotify returns a channel which will be closed (i.e. could
+// be read from) as soon as the test execution is resumed.
+//
+// Since tests would likely be paused only rarely, unless you
+// directly need to be notified via a channel that the test
+// isn't paused or that it has resumed, it's probably a good
+// idea to first use the IsPaused() method, since it will be much
+// faster.
+//
+// And, since tests won't be paused most of the time, it's
+// probably better to check for that like this:
+//   if executionState.IsPaused() {
+//       <-executionState.ResumeNotify()
+//   }
+func (es *ExecutionState) ResumeNotify() <-chan struct{} {
+	es.pauseStateLock.RLock()
+	defer es.pauseStateLock.RUnlock()
+	return es.resumeNotify
+}
+
+// GetPlannedVU tries to get a pre-initialized VU from the buffer channel. This
+// shouldn't fail and should generally be an instantaneous action, but if it
+// doesn't happen for MaxTimeToWaitForPlannedVU, a warning will be printed. If
+// we reach that timeout more than MaxRetriesGetPlannedVU number of times, this
+// function will return an error, since we either have a bug with some
+// executor, or the machine is very, very overloaded.
+func (es *ExecutionState) GetPlannedVU(logger *logrus.Entry, modifyActiveVUCount bool) (VU, error) {
```

Comment:
Minor: could you document the purpose of `modifyActiveVUCount`? It's currently not very clear why the externally-controlled executor is the only one that passes this as `false` (explaining it there as well wouldn't hurt).

Reply:
Resolved by https://github.com/loadimpact/k6/pull/1007/commits/61069240e34c89dcfa5464648e1caab2cd857046